testing/e2e/infra/provision.go

```diff
@@ -0,0 +1,230 @@
+
+	for idx := 0; idx < lenZones; idx++ {
+		func(idx int) {
+			resEg.Go(func() error {
+				zone, err := clients.NewZone(ctx, config.Flags.SubscriptionId, i.ResourceGroup, fmt.Sprintf("zone-%d-%s", idx, i.Suffix))
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating zone: %w", err))
+				}
+				ret.Zones = append(ret.Zones, zone)
+				return nil
+			})
+		}(idx)
+	}
+	for idx := 0; idx < lenPrivateZones; idx++ {
+		func(idx int) {
+			resEg.Go(func() error {
+				privateZone, err := clients.NewPrivateZone(ctx, config.Flags.SubscriptionId, i.ResourceGroup, fmt.Sprintf("private-zone-%d-%s", idx, i.Suffix))
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating private zone: %w", err))
+				}
+				ret.PrivateZones = append(ret.PrivateZones, privateZone)
+				return nil
+			})
+		}(idx)
+	}
+
+	if err := resEg.Wait(); err != nil {
+		return ProvisionedInfra{}, logger.Error(lgr, err)
+	}
+
+	// connect permissions
+	var permEg errgroup.Group
+
+	for _, pz := range ret.PrivateZones {
+		func(pz privateZone) {
+			permEg.Go(func() error {
+				cluster, err := ret.Cluster.GetCluster(ctx)
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("getting cluster: %w", err))
+				}
+
+				dns, err := pz.GetDns(ctx)
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("getting dns: %w", err))
+				}
+
+				principalId := cluster.Identity.PrincipalID
+				role := clients.PrivateDnsContributorRole
+				if _, err := clients.NewRoleAssignment(ctx, config.Flags.SubscriptionId, *dns.ID, *principalId, role); err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating %s role assignment: %w", role.Name, err))
+				}
+
+				vnet, err := ret.Cluster.GetVnetId(ctx)
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("getting vnet id: %w", err))
+				}
+				if err := pz.LinkVnet(ctx, fmt.Sprintf("link-%s-%s", pz.GetName(), i.Suffix), vnet); err != nil {
+					return logger.Error(lgr, fmt.Errorf("linking vnet: %w", err))
+				}
+
+				return nil
+			})
+		}(pz)
+	}
+
+	for _, z := range ret.Zones {
+		func(z zone) {
+			permEg.Go(func() error {
+				cluster, err := ret.Cluster.GetCluster(ctx)
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("getting cluster: %w", err))
+				}
+
+				dns, err := z.GetDns(ctx)
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("getting dns: %w", err))
+				}
+
+				principalId := cluster.Identity.PrincipalID
+				role := clients.DnsContributorRole
+				if _, err := clients.NewRoleAssignment(ctx, config.Flags.SubscriptionId, *dns.ID, *principalId, role); err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating %s role assignment: %w", role.Name, err))
+				}
+
+				return nil
+			})
+		}(z)
+	}
+
+	permEg.Go(func() error {
+		cluster, err := ret.Cluster.GetCluster(ctx)
+		if err != nil {
+			return logger.Error(lgr, fmt.Errorf("getting cluster: %w", err))
+		}
+
+		kubelet, ok := cluster.Properties.IdentityProfile["kubeletidentity"]
```

Comment:
rather than use the kubelet identity, we are better off using the app routing specific identity - that way our e2e will be more similar to the user's case. should be in [ingressprofile.webapprouting.identity](https://github.com/Azure/azure-rest-api-specs/blob/033e403776fbd3ee48503b87f27dce527324d70a/specification/containerservice/resource-manager/Microsoft.ContainerService/aks/preview/2023-04-02-preview/managedClusters.json#L6843)

Reply:
This is the AcrPull role. AcrPull does go on the kubelet.

We also don't have access to the app routing identity because we are not using the addon in these tests, simply replicating it.

Reply:
very true